core/arch/riscv64/riscv64.asm

```diff
@@ -233,7 +233,10 @@ GLOBAL_LABEL(dr_longjmp:)
         /* int atomic_swap(int *adr, int val) */
         DECLARE_FUNC(atomic_swap)
 GLOBAL_LABEL(atomic_swap:)
-/* FIXME i#3544: Not implemented */
+1:      lr.d     ARG3, (ARG1)
+        sc.d     ARG4, ARG2, (ARG1)
+        bnez     ARG4, 1b
+        mv       ARG1, ARG3
```

Comment:
Just use `amoswap.d`.

Reply:
Maybe amoswap.d has a better performance. But the instruction is too long.
It looks like this:
```
         amoswap.d     ARG1, ARG2, (ARG1)
1:      lr.d     ARG3, (ARG1)
        sc.d     ARG4, ARG2, (ARG1)
        bnez     ARG4, 1b
        mv       ARG1, ARG3
```

Reply:
> But the instruction is too long.

I don't understand the concern. With `amoswap` it takes just one instruction to accomplish the `atomic_swap`. 

Also, in your snippet I see `amoswap` and the lr-sc sequence both; however we just need one of them right?

Reply:
> Also, in your snippet I see amoswap and the lr-sc sequence both;
Only one is needed.

>  With amoswap it takes just one instruction to accomplish the atomic_swap.
These two options are both viable,  amoswap has better readability. Here, I would like to discuss whether the use of amoswap exceeds Dynamorio's coding style conventions(In .asm files, place opcodes on column 8 and start operands on column 17.). Maybe we should make an exception?

Reply:
Yes, it should be fine.